## Classification (*Week 2*)
### Answers Briefly (*Details shown in following pages*)
* 
#### Q5:
* Accuracy on training set is 91.296
* Accuracy on test set is 92.112
#### Q6:
* feature: whether there is ' IPA ' in the 'beer/name'
* Accuracy on training set is 96.192
* Accuracy on test set is 93.74
#### Q7:
* Accuracy on training set is always 96.168 
* Accuracy on test set is always 93.788
* if using RBF kernel, accuracy changes

In [1]:
import numpy
import scipy.optimize
import random
from sklearn import svm

def parseDataFromFile(fname):
    for l in open(fname):
        yield eval(l)

print "Reading data..."
data = list(parseDataFromFile("beer_50000.json"))
print "done"
data[14712] #show a sample of data

Reading data...
done


{'beer/ABV': 6.8,
 'beer/beerId': '505',
 'beer/brewerId': '14',
 'beer/name': 'Big Man Ale',
 'beer/style': 'American IPA',
 'review/appearance': 3.0,
 'review/aroma': 3.5,
 'review/overall': 2.5,
 'review/palate': 3.5,
 'review/taste': 3.0,
 'review/text': '"He\'s a big, strong man with a child\'s mind, don\'t you take his booze away!" - Dropkick Murphys\t\t@ Atwoods, Sommerville MA. Found an extra $20 that I didn\'t know I had so I took my ladyfriend out for a beer.\t\tHazy copper color with a thin white head. Momentary thin skim throws up a few lacey splashes before dying out. Honestly, it wasn\'t a great pour.\t\tFloral and citrus hop aroma with bubble gum. There\'s a very faint, mild fruit malt under there somewhere.\t\tI didn\'t get the citrus bomb that many reviewers described but rather a clean, juniper and spruce bitter. Bubble gum pops up here and there with a piney caramel fade. Hoppy burps that tastes like I swallowed a pine tree. \t\tOdd malt character: I can feel it but 

---
## Qestion 5: 
#### c train a predictor that estimates whether a beer is an ‘American IPA’ using two features:
#### *[‘beer/ABV’, ‘review/taste’]*
#### Train your predictor using an SVM classifier, What is the accuracy

In [2]:
#Building feature X and label Y
X=[]
Y=[0]*50000
for i in range(50000):
    if data[i]['beer/style']== 'American IPA':
        Y[i]=1
    X.append([data[i]['beer/ABV'],data[i]['review/taste']])

#spliting dataset
X_train = X[:25000]
y_train = Y[:25000]
X_test = X[25000:]
y_test = Y[25000:]

#training data with Linear Kernel SVM
clf = svm.SVC(C=1000,kernel='linear')
clf.fit(X_train, y_train)
train_predictions = clf.predict(X_train)
test_predictions = clf.predict(X_test)

#calculating accuracy
train_accuracy=sum([(x==y) for x,y in zip(train_predictions,y_train)])*1.0/25000*100
test_accuracy=sum([(x==y) for x,y in zip(test_predictions,y_test)])*1.0/25000*100

In [3]:
print train_accuracy,test_accuracy

91.296 92.112


### training data with Linear Kernel SVM:
* Accuracy on training set is 91.296
* Accuracy on test set is 92.112

---
## Qestion 6: 
#### a more accurate predictor：
#### The feature I choose: whether there is an  ** ' IPA ' ** in  the **'beer/name'**


In [4]:
X=[]
for i in range(50000):
    if 'IPA' in data[i]['beer/name']:
        X.append([1,data[i]['review/taste']])
    else:
        X.append([0,data[i]['review/taste']])

X_train = X[25000:]
y_train = Y[25000:]
X_test = X[:25000]
y_test = Y[:25000]

clf = svm.SVC(C=1000)
clf.fit(X_train, y_train)
train_predictions = clf.predict(X_train)
train_accuracy=sum([(x==y) for x,y in zip(train_predictions,y_train)])*1.0/25000*100
test_predictions = clf.predict(X_test)
test_accuracy=sum([(x==y) for x,y in zip(test_predictions,y_test)])*1.0/25000*100

In [5]:
print train_accuracy,test_accuracy

96.192 93.74


### Using the new feature:
* Accuracy on training set is 96.192
* Accuracy on test set is 93.74
* BTW: The RBF kernel significantly speed up the modeling

---
## Qestion 7: 
#### What effect does the regularization constant C have on the training/test performance?

In [7]:
clf = svm.SVC(C=100000,kernel='linear')
clf.fit(X_train, y_train)
train_predictions = clf.predict(X_train)
train_accuracy=sum([(x==y) for x,y in zip(train_predictions,y_train)])*1.0/25000*100
test_predictions = clf.predict(X_test)
test_accuracy=sum([(x==y) for x,y in zip(test_predictions,y_test)])*1.0/25000*100
print train_accuracy,
print test_accuracy

96.168 93.788


In [8]:
clf = svm.SVC(C=10,kernel='linear')
clf.fit(X_train, y_train)
train_predictions = clf.predict(X_train)
train_accuracy=sum([(x==y) for x,y in zip(train_predictions,y_train)])*1.0/25000*100
test_predictions = clf.predict(X_test)
test_accuracy=sum([(x==y) for x,y in zip(test_predictions,y_test)])*1.0/25000*100
print train_accuracy,
print test_accuracy

96.168 93.788


In [9]:
clf = svm.SVC(C=0.1,kernel='linear')
clf.fit(X_train, y_train)
train_predictions = clf.predict(X_train)
train_accuracy=sum([(x==y) for x,y in zip(train_predictions,y_train)])*1.0/25000*100
test_predictions = clf.predict(X_test)
test_accuracy=sum([(x==y) for x,y in zip(test_predictions,y_test)])*1.0/25000*100
print train_accuracy,
print test_accuracy

96.168 93.788


### Using the C as 100000, 10 , 0.1, 1000 with **Linear Kernel**:
* Accuracy on training set is always 96.168 
* Accuracy on test set is always 93.788
### However, if changed into RBF Kernel: (they do not remain the same)
* C=1000: Train=96.2; Test=93.7;
* C=100000: Train=96.2; Test=93.7;
* C=10: Train=96.2; Test=93.8;
* C=1000: Train=96.2; Test=93.8;